File: frontend/src/lib/sequencerTimingMigration.ts
/** Shared migration contract: backend/tests/fixtures/sequencers/timing_migration.json. */
type RecordValue = Record<string, unknown>;
const record = (value: unknown): value is RecordValue => !!value && typeof value === "object" && !Array.isArray(value);

function doubleLengths(value: unknown): unknown {
  if (Array.isArray(value)) return value.map(doubleLengths);
  if (!record(value)) return value;
  return Object.fromEntries(Object.entries(value).map(([key, item]) => {
    if ((key === "lengthBeats" || key === "length_beats") && typeof item === "number") return [key, item * 2];
    if (key === "lengths" && Array.isArray(item)) return [key, item.map(n => Number(n) * 2)];
    if ((key === "padLoopSequence" || key === "pad_loop_sequence") && Array.isArray(item)) {
      return [key, item.map(n => typeof n === "number" && n < 0 ? n * 2 : n)];
    }
    return [key, doubleLengths(item)];
  }));
}

function canonicalLengths(value: unknown, kind: string): unknown {
  if (Array.isArray(value)) return value.map(item => canonicalLengths(item, kind));
  if (!record(value)) return value;
  const result = Object.fromEntries(Object.entries(value).map(([key, item]) => [key, canonicalLengths(item, kind)]));
  if (result.lengthBeats === undefined && result.length_beats === undefined && (result.stepCount !== undefined || result.step_count !== undefined)) {
    let length = result.stepCount ?? result.step_count;
    if (kind !== "controller" && length === 16) length = 4;
    else if (length === 32) length = 8;
    else if (length === 64) length = kind === "controller" ? 16 : 4;
    else if (typeof length !== "number" || !((length >= 1 && length <= 8) || kind === "controller" && length === 16)) length = 4;
    result.lengthBeats = length;
  }
  return result;
}

export function migrateSequencerTiming(raw: unknown, appState = false): unknown {
  const targetVersion = appState ? 3 : 17;
  if (!record(raw) || !Number.isInteger(raw.version ?? 1) || Number(raw.version ?? 1) < 1 || Number(raw.version ?? 1) >= targetVersion || !record(raw.sequencer)) return raw;
  const result = structuredClone(raw);
  const sequencer = result.sequencer as RecordValue;
  const keys = [["tempoBPM", "tempo_bpm"], ["meterNumerator", "meter_numerator"], ["meterDenominator", "meter_denominator"], ["stepsPerBeat", "steps_per_beat"], ["beatRateNumerator", "beat_rate_numerator"], ["beatRateDenominator", "beat_rate_denominator"]];
  const globalTiming = record(sequencer.timing) ? sequencer.timing : Object.fromEntries(keys.map(([key, alias]) => [key, sequencer[key] ?? sequencer[alias]]).filter(([, value]) => value !== undefined));
  const migrated = new Set<string>();
  for (const [field, kind] of [["tracks", "sequencer"], ["drummerTracks", "drummer"], ["controllerSequencers", "controller"]]) {
    const tracks = sequencer[field];
    if (!Array.isArray(tracks)) continue;
    sequencer[field] = tracks.map(rawTrack => {
      if (!record(rawTrack)) return rawTrack;
      rawTrack = canonicalLengths(rawTrack, kind) as RecordValue;
      const timing = record(rawTrack.timing) ? rawTrack.timing : Object.fromEntries(keys.map(([key, alias]) =>
        [key, rawTrack[key] ?? rawTrack[alias] ?? globalTiming[key] ?? globalTiming[alias]]).filter(([, value]) => value !== undefined));
      if ((timing.meterDenominator ?? timing.meter_denominator ?? 4) !== 8) return rawTrack;
      const track = doubleLengths({ ...rawTrack, lengthBeats: rawTrack.lengthBeats ?? rawTrack.length_beats ?? 4 }) as RecordValue;
      const nextTiming: RecordValue = { ...timing, stepsPerBeat: Number(timing.stepsPerBeat ?? timing.steps_per_beat ?? 4) / 2 };
      delete nextTiming.steps_per_beat;
      track.timing = nextTiming;
      migrated.add(`${kind}:${rawTrack.id ?? rawTrack.trackId}`);
      return track;
    });
  }
  const history = result.arrangerHistory;
  const convert = (item: unknown) => record(item) && migrated.has(`${item.kind}:${item.id}`) ? doubleLengths(canonicalLengths(item, String(item.kind))) : item;
  if (record(history)) {
    if (Array.isArray(history.basis)) history.basis = history.basis.map(item => { const next = convert(item); return next !== item && record(next) ? { ...next, meterDenominator: 8 } : next; });
    if (Array.isArray(history.entries)) for (const entry of history.entries) {
      if (record(entry) && Array.isArray(entry.changes)) entry.changes = entry.changes.map(convert);
    }
  }
  result.version = targetVersion;
  return result;
}

File: backend/app/services/sequencer_timing_migration.py
"""Versioned, non-destructive migration from quarter beats to meter beats.

Keep this contract aligned with frontend/lib/sequencerTimingMigration.ts and
the standalone performance CLI; shared fixtures exercise all three readers.
"""
from copy import deepcopy


def _double_lengths(value):
    if isinstance(value, list):
        return [_double_lengths(item) for item in value]
    if not isinstance(value, dict):
        return value
    result = {}
    for key, item in value.items():
        if key in ("lengthBeats", "length_beats") and isinstance(item, (int, float)):
            result[key] = item * 2
        elif key == "lengths" and isinstance(item, list):
            result[key] = [length * 2 for length in item]
        elif key in ("padLoopSequence", "pad_loop_sequence") and isinstance(item, list):
            result[key] = [token * 2 if isinstance(token, int) and token < 0 else token for token in item]
        else:
            result[key] = _double_lengths(item)
    return result


def _canonical_lengths(value, kind):
    if isinstance(value, list):
        return [_canonical_lengths(item, kind) for item in value]
    if not isinstance(value, dict):
        return value
    result = {key: _canonical_lengths(item, kind) for key, item in value.items()}
    if not any(key in result for key in ("lengthBeats", "length_beats")) and any(key in result for key in ("stepCount", "step_count")):
        length = result.get("stepCount", result.get("step_count"))
        if kind != "controller" and length == 16:
            length = 4
        elif length == 32:
            length = 8
        elif length == 64:
            length = 16 if kind == "controller" else 4
        elif not isinstance(length, (int, float)) or not (1 <= length <= 8 or kind == "controller" and length == 16):
            length = 4
        result["lengthBeats"] = length
    return result


def migrate_sequencer_timing(config: dict, *, app_state: bool = False) -> dict:
    target_version = 3 if app_state else 17
    if not isinstance(config, dict) or type(config.get("version", 1)) is not int or config.get("version", 1) not in range(1, target_version):
        return config
    sequencer = config.get("sequencer")
    if not isinstance(sequencer, dict):
        return config
    result = deepcopy(config)
    sequencer = result["sequencer"]
    timing_keys = (("tempoBPM", "tempo_bpm"), ("meterNumerator", "meter_numerator"), ("meterDenominator", "meter_denominator"), ("stepsPerBeat", "steps_per_beat"), ("beatRateNumerator", "beat_rate_numerator"), ("beatRateDenominator", "beat_rate_denominator"))
    global_timing = sequencer["timing"] if isinstance(sequencer.get("timing"), dict) else {key: sequencer.get(key, sequencer.get(alias)) for key, alias in timing_keys if key in sequencer or alias in sequencer}
    migrated = set()
    for field, kind in (("tracks", "sequencer"), ("drummerTracks", "drummer"), ("controllerSequencers", "controller")):
        tracks = sequencer.get(field)
        if not isinstance(tracks, list):
            continue
        for index, raw_track in enumerate(tracks):
            if not isinstance(raw_track, dict):
                continue
            track = _canonical_lengths(raw_track, kind)
            timing = track["timing"] if isinstance(track.get("timing"), dict) else {key: track.get(key, track.get(alias, global_timing.get(key, global_timing.get(alias))))
                for key, alias in timing_keys if key in track or alias in track or key in global_timing or alias in global_timing}
            denominator = timing.get("meterDenominator", timing.get("meter_denominator", 4))
            sequencer[field][index] = track
            if denominator != 8:
                continue
            track.setdefault("lengthBeats", track.get("length_beats", 4))
            converted = _double_lengths(track)
            converted["timing"] = deepcopy(timing)
            grid = timing.get("stepsPerBeat", timing.get("steps_per_beat", 4))
            converted["timing"]["stepsPerBeat"] = grid // 2
            converted["timing"].pop("steps_per_beat", None)
            sequencer[field][index] = converted
            identity = track.get("id", track.get("trackId"))
            if isinstance(identity, str):
                migrated.add((kind, identity))
    history = result.get("arrangerHistory")
    def convert(item):
        if isinstance(item, dict) and isinstance(item.get("kind"), str) and isinstance(item.get("id"), str) and (item["kind"], item["id"]) in migrated:
            return _double_lengths(_canonical_lengths(item, item.get("kind")))
        return item

    if isinstance(history, dict):
        if isinstance(history.get("basis"), list):
            basis_values = []
            for basis in history["basis"]:
                converted = convert(basis)
                basis_values.append({**converted, "meterDenominator": 8} if converted is not basis else basis)
            history["basis"] = basis_values
        if isinstance(history.get("entries"), list):
            for entry in history["entries"]:
                if isinstance(entry, dict) and isinstance(entry.get("changes"), list):
                    entry["changes"] = [convert(change) for change in entry["changes"]]
    result["version"] = target_version
    return result
